# Udacity Machine Learning Nanodegree

## Capstone Project — Tweet Classifier

> *Classify tweets as either __Republican__ or __Democrat__.*

The goal of this project is to create a neural network text classifier that performs better then a benchmark Naive Bayes classifier.

---

## Step 1 — Examine the Dataset

Some summary stats:

In [137]:
from sklearn.datasets import load_files
import pandas as pd
import numpy as np

tweets = pd.read_csv('./dataset/ExtractedTweets.csv')
#tweets.head()
print "Total number of tweets: %i" % tweets['Tweet'].count()

print "------"
party_count = tweets.groupby(['Party',]).count()
print ("Tweets by party:")
print(party_count['Tweet'])

print "------"
handles = None
handle_counts = None
handles = tweets.groupby('Handle')
handle_counts = handles.count()
non_two_hundred = handle_counts[handle_counts['Tweet']!=200]['Tweet'].count()
print "Mean number tweets per account: %f" % handle_counts['Tweet'].mean()
print "Median number of tweets per account: %f" % handle_counts['Tweet'].median()
print "Number of accounts without exactly two hundred tweets: %f" % non_two_hundred


Total number of tweets: 86460
------
Tweets by party:
Party
Democrat      42068
Republican    44392
Name: Tweet, dtype: int64
------
Mean number tweets per account: 199.676674
Median number of tweets per account: 200.000000
Number of accounts without exactly two hundred tweets: 17.000000


This is an almost perfectly balanced binary-labeled dataset. There are a total of **86460** tweets.  **42068** (48.65%) are labeled *Democrat*, while **44392** (51.34%) are labeled *Republican*. All but seventeen (17) of the accounts included in this data set made exactly two-hundred (200) tweets each.

This data set is ideal for doing a binary classification based solely on the content of the text in the dataset, with minimal influence from data composition.

## Step 2 — Vectorize

---


In [153]:
from sklearn.feature_extraction.text import CountVectorizer
count_vect = CountVectorizer()
X_train_counts = count_vect.fit_transform(tweets['Tweet'])
X_train_counts.shape


(86460, 126329)